File: tests/UITests/PageObjectModels/TheInternetPages/FloatingMenuPage.cs
using Microsoft.Playwright;

namespace Tests.UITests.PageObjectModels.TheInternetPages;

public class FloatingMenuPage : BasePage
{
    private const string Url = "https://the-internet.herokuapp.com/floating_menu";

    private ILocator HomeLink => Page.GetByRole(AriaRole.Link, new() { Name = "Home" });
    private ILocator NewsLink => Page.GetByRole(AriaRole.Link, new() { Name = "News" });
    private ILocator ContactLink => Page.GetByRole(AriaRole.Link, new() { Name = "Contact" });
    private ILocator AboutLink => Page.GetByRole(AriaRole.Link, new() { Name = "About" });

    public FloatingMenuPage(IPage page) : base(page) { }

    public async Task Navigate() => await NavigateTo(Url);

    public async Task<bool> IsMenuVisible()
    {
        return await HomeLink.IsVisibleAsync()
            && await NewsLink.IsVisibleAsync()
            && await ContactLink.IsVisibleAsync()
            && await AboutLink.IsVisibleAsync();
    }

    public async Task ClickNavLink(string name) =>
        await Page.GetByRole(AriaRole.Link, new() { Name = name }).ClickAsync();

    public async Task<bool> IsMenuVisibleAfterScroll()
    {
        await Page.EvaluateAsync("window.scrollTo(0, document.body.scrollHeight)");
        return await IsMenuVisible();
    }
}

File: tests/UITests/PageObjectModels/TheInternetPages/FloatingMenuPage.js
const { BasePage } = require('./BasePage');

class FloatingMenuPage extends BasePage {
  constructor(page) {
    super(page);
    this.url = '/floating_menu';
    this.homeLink = page.getByRole('link', { name: 'Home' });
    this.newsLink = page.getByRole('link', { name: 'News' });
    this.contactLink = page.getByRole('link', { name: 'Contact' });
    this.aboutLink = page.getByRole('link', { name: 'About' });
  }

  async navigate() {
    await this.navigateTo(this.url);
  }

  async isMenuVisible() {
    return (
      (await this.homeLink.isVisible()) &&
      (await this.newsLink.isVisible()) &&
      (await this.contactLink.isVisible()) &&
      (await this.aboutLink.isVisible())
    );
  }

  async clickNavLink(name) {
    await this.page.getByRole('link', { name }).click();
  }

  async isMenuVisibleAfterScroll() {
    await this.page.evaluate(() => window.scrollTo(0, document.body.scrollHeight));
    return this.isMenuVisible();
  }
}

module.exports = { FloatingMenuPage };
